File: templates/flutter/owl_flutter/lib/utils/owl.dart
import 'dart:convert';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:owl_flutter/components/owl_home.dart';
import 'package:owl_flutter/owl_generated/owl_route.dart';

import 'package:owl_flutter/pageJump/page_jumpper.dart';


import '../owl_generated/owl_app.dart';

class Owl {
  BuildContext docBuildContext;
  WeiXinAdapter wx;

  void login(app) {}

  void getUserInfo(app, page) {}

  bool canIUse(buttonName) {
    return true;
  }

  static bool isHomeTabUrl(url) {
    var tabBar = getApplication().appJson['tabBar'];
    if (tabBar == null) {
      return false;
    }

    var list = tabBar['list'];

    if (list == null) {
      return false;
    }
    for (int i = 0; i < list.length; i++) {
      var item = list[i];
      if (url == item['pagePath']) {
        return true;
      }
    }
    return false;
  }

  static OwlApp owlapp = OwlApp();

  static OwlApp getApplication() {
    return owlapp;
  }

  static void navigateTo(var obj, BuildContext context) {
    String url = obj['url'];
    if (url.startsWith("/")) {
      url = url.substring(1);
    }

    int pos = url.indexOf("?");
    Map params = {};
    if (pos > 0) {
      String queryString = url.substring(pos + 1);
      String path = url.substring(0, pos);
      url = path;
      //parse queryString

      List<String> paramPairs = queryString.split("&");
      for (int i = 0; i < paramPairs.length; i++) {
        String paramPair = paramPairs[i];
        List<String> kv = paramPair.split("=");
        if (kv.length == 2) {
          params[kv[0]] = kv[1];
        }
      }
    }
    //检查url是否属于tabs, 如果属于则不跳转
    var tabBar = getApplication().appJson['tabBar'];
    if (tabBar != null) {
      var list = tabBar['list'];

      for (int i = 0; i < list.length; i++) {
        var tab = list[i];
        var pagePath = tab['pagePath'];
        if (pagePath == url) {
          print(pagePath + " is one of the tab, use wx.switchTab instead");
          return null;
        }
      }
    }

    Map  map = Map();
    map["pageType"] = "flutter/"+url;
    //map["naviBarHidden"] ='1';
    PageJumpper.notityNativePush(map,context: context);


//    print("navigating to $url, params=$params");
//    Widget screen = getScreen(url, params, owl.getApplication().appCss);
//    if (screen != null) {
//      Navigator.push(
//        context,
//        MaterialPageRoute(builder: (context) => screen),
//      );
//    }
  }

  static void switchTab(var obj, BuildContext context) {
    String url = obj['url'];
    if (url.startsWith("/")) {
      url = url.substring(1);
    }

    //判断url是否属于某个tab
    var tabBar = getApplication().appJson['tabBar'];
    if (tabBar == null) {
      return null;
    }

    var list = tabBar['list'];
    if (list == null) {
      return null;
    }
    var currentIndex = -1;

    for (int i = 0; i < list.length; i++) {
      var item = list[i];
      if (item['pagePath'] == url) {
        currentIndex = i;
      }
    }
    if (currentIndex == -1) {
      return null;
    }

    int pos = url.indexOf("?");
    Map params = {};
    if (pos > 0) {
      String queryString = url.substring(pos + 1);
      String path = url.substring(0, pos);
      url = path;
      //parse queryString

      List<String> paramPairs = queryString.split("&");
      for (int i = 0; i < paramPairs.length; i++) {
        String paramPair = paramPairs[i];
        List<String> kv = paramPair.split("=");
        if (kv.length == 2) {
          params[kv[0]] = kv[1];
        }
      }
    }

    Widget screen = OwlHome(url, params);
    if (screen != null) {
      PageJumpper.notityNativeReplacementPush((context) => screen);

    }
  }

  static void navigateBack(var obj, BuildContext context) {
    int delta = obj['delta'];
    for (int i = 0; i < delta; i++) {
//      Navigator.pop(context);
        PageJumpper.notityNativePop();
    }
  }

  void addToList(arr, elem) {
    arr.add(elem);
  }

  void removeFromList(arr, index) {
    arr.removeAt(index);
  }

  void insert(arr, index, elem) {
    arr.insert(index, elem);
  }

  bool containsKey(map, String key) {
    return map.containsKey(key);
  }

  void uploadFiles(url, pictures, onSuccess, onFailed) {
    var uploaded = 0;
    var failed = false;

    var total = pictures.length;
    List uploadedPictures = new List(total);
    wx.showLoading({
      "title": "上传中:" + uploaded.toString() + "/" + total.toString(),
      "mask": true
    });
    for (var i = 0; i < pictures.length; i++) {
      var item = pictures[i];
      wx.uploadFile({
        "url": url,
        "filePath": item,
        "name": "pic",
        "formData": {},
        "success": (res) {
          //console.log(res);
          uploaded = uploaded + 1;
          print("uploaded index:$i");
          uploadedPictures[i] = json.decode(res["data"]);
          wx.showLoading({
            "title": "上传中:" + uploaded.toString() + "/" + total.toString(),
            "mask": true
          });
          if (uploaded == total) {
            wx.hideLoading({});
            wx.showToast({"title": "上传成功。", "icon": "success"});
            if (onSuccess != null) {
              if (!failed) {
                onSuccess(uploadedPictures);
              }
            }
          }
        },
        "fail": (e) {
          wx.hideLoading({});
          wx.showToast({"title": "上传出错"});
          if (onFailed != null && !failed) {
            failed = true;
            onFailed({"index": i});
          }
        }
      });
    }
  }

  String jsonEncode(Object v) {
    return json.encode(v);
  }

  dynamic jsonDecode(String source) {
    return json.decode(source);
  }
}

File: templates/wechat/owl_wechat/utils/owl.js

function login(app) {
    wx.login({
        success: res => {
            // 发送 res.code 到后台换取 openId, sessionKey, unionId
            console.log("login return:",res);
        }
    })
}

function getUserInfo(app,page) {
    if (app.globalData.userInfo) {
        page.setData({
            userInfo: app.globalData.userInfo,
            hasUserInfo: true
        })
        return;
    }
    if(!page){
        //在app.js中调用
        wx.getSetting({
            success: res => {
                console.log('owlGetUserInfo success',res);
                if (res.authSetting['scope.userInfo']) {
                    // 已经授权，可以直接调用 getUserInfo 获取头像昵称，不会弹框
                    wx.getUserInfo({
                        success: res => {
                            // 可以将 res 发送给后台解码出 unionId
                            app.globalData.userInfo = res.userInfo
                            // 由于 getUserInfo 是网络请求，可能会在 Page.onLoad 之后才返回
                            // 所以此处加入 callback 以防止这种情况
                            if (app.userInfoReadyCallback) {
                                app.userInfoReadyCallback(res)
                            }
                        }
                    })
                }
            }
        })
    }
    else{
        //在page中调用
        var canIUse = wx.canIUse('button.open-type.getUserInfo')
        page.setData({canIUse:canIUse});
        if (canIUse){
            // 由于 getUserInfo 是网络请求，可能会在 Page.onLoad 之后才返回
            // 所以此处加入 callback 以防止这种情况
            app.userInfoReadyCallback = res => {
                page.setData({
                    userInfo: res.userInfo,
                    hasUserInfo: true
                })
            }
        } else {
            // 在没有 open-type=getUserInfo 版本的兼容处理
            wx.getUserInfo({
                success: res => {
                    app.globalData.userInfo = res.userInfo
                    page.setData({
                        userInfo: res.userInfo,
                        hasUserInfo: true
                    })
                }
            })
        }
    }

}

function canIUse(buttonName){
    //wx.canIUse('button.open-type.getUserInfo')
    return wx.canIUse(buttonName);
}

function navigateTo(url){
    return wx.navigateTo({url:url});
}

function getApplication(){
    return getApp(); //wechat api
}

function addToList(arr,elem){
    arr.push(elem);
}


function removeFromList(arr, index){
    arr.splice(arr,index,1);
}

function insert(arr,index,elem){
    arr.splice(arr,index,0,elem);
}

module.exports = {
    login,
    getUserInfo,
    canIUse,
    navigateTo,
    getApplication
}
